### PR TITLE
added support to spring amqp batch

### LINKED ISSUE #1706: Extending Spring AMQP support for message batch API
## Is your feature request related to a problem?
Message batch processing API of Spring AMQP is not yet supported - `org.springframework.amqp.core.MessageListener#onMessageBatch(List<Message> messages)` 

## Describe the solution you'd like
If possible, wrapping the list of messages provided as argument with one that uses the list iteration to create and manage transactions. Similar to the [Kafka](https://github.com/elastic/apm-agent-java/blob/master/apm-agent-plugins/apm-kafka-plugin/apm-kafka-headers-plugin/src/main/java/co/elastic/apm/agent/kafka/helper/ConsumerRecordsIteratorWrapper.java) approach.

## Describe alternatives you've considered
Wait for span links and create one transaction for the batch, instead of transaction-per-message.

## Additional context
A followup on #1657 
@kananindzya If you already have a test for that, you can start a new draft PR based on it.

### PR DESCRIPTION
**Pull request by kananindzya**
closes #1706
## Checklist

- [X] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [X] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/MessageBatchIteratorWrapper.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+        Message message = delegate.next();
+        try {
+            if (message == null) {
+                return message;
+            }
+            MessageProperties messageProperties = message.getMessageProperties();
+            if (messageProperties == null) {
+                return message;
+            }
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/MessageBatchIteratorWrapper.java:57]: Why? We should create a transaction anyway - a root transaction if there is no `tracecontext` property, or child transaction if there is. We already do that (create a transaction when there is no property). The same goes for `SpringAmqpMessageListenerInstrumentation` - we should allow `null` properties.
Please fix `SpringAmqpTransactionHelper#createTransaction()` instead to allow for `null` properties (and treat the same as if there are properties without context property).

**Review by eyalkoren — COMMENTED** [review]
Nice! 👏 
A few minor comments, overall looks really good!
I am pushing a small fix that will allow testing.

File: apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/MessageBatchIteratorWrapper.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    @Override
+    public Message next() {
+        endCurrentTransaction();
+
+        Message message = delegate.next();
+        try {
+            if (message == null) {
+                return message;
+            }
```
[the remark this answers is not in the document]
Comment (kananindzya) [reply apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/MessageBatchIteratorWrapper.java:53]: deleted

File: apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/MessageBatchHelper.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package co.elastic.apm.agent.rabbitmq;
+
+import java.util.List;
+
+public interface MessageBatchHelper<MSG> {
+
+    List<MSG> wrapMessageBatchList(List<MSG> messageBatchList);
+}
```
[the remark this answers is not in the document]
Comment (kananindzya) [reply apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/MessageBatchHelper.java:8]: deleted

File: apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/SpringBaseInstrumentation.java [hunk 4]
```diff
@@ -45,4 +56,15 @@
         return not(isBootstrapClassLoader())
             .and(classLoaderCanLoadClass("org.springframework.amqp.core.MessageListener"));
     }
+
+    static class BaseAdvice {
+        protected static final MessageBatchHelper messageBatchHelper;
```
[the remark this answers is not in the document]
Comment (kananindzya) [reply apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/SpringBaseInstrumentation.java:61]: moved to SpringAmqpBatchMessageListenerInstrumentation$MessageListenerContainerWrappingAdvice

File: apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/SpringAmqpTransactionHelper.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package co.elastic.apm.agent.rabbitmq;
+
+import co.elastic.apm.agent.impl.transaction.Transaction;
+import org.springframework.amqp.core.Message;
+import org.springframework.amqp.core.MessageProperties;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public interface SpringAmqpTransactionHelper {
```
[the remark this answers is not in the document]
Comment (kananindzya) [reply apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/SpringAmqpTransactionHelper.java:10]: yes, this interface not required. Deleted

File: apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/MessageBatchIteratorWrapper.java (hunk 1 above)
Comment (kananindzya) [reply apm-agent-plugins/apm-rabbitmq/apm-rabbitmq-spring/src/main/java/co/elastic/apm/agent/rabbitmq/MessageBatchIteratorWrapper.java]: fixed

**Comment by eyalkoren** [comment]
Thanks for applying the review comments ❤️ 

Please see my last commits - I only removed another interface, changed helper names, added to CHANGELOG and added a test for distributed tracing.

Please let me know if you disagree with any of this.

**Review by eyalkoren — APPROVED** [review]

**Comment by eyalkoren** [comment]
run elasticsearch-ci/docs

**Comment by eyalkoren** [comment]
@kananindzya we removed our `AssignTo` annotations and migrated to use the new Byte Buddy `AssignReturned` annotations - see other instrumentations and please it can build and test pass.
Sorry about that 🙁

**Comment by kananindzya** [comment]
> @kananindzya we removed our `AssignTo` annotations and migrated to use the new Byte Buddy `AssignReturned` annotations - see other instrumentations and please it can build and test pass. Sorry about that slightly_frowning_face

fixed

**Comment by eyalkoren** [comment]
run elasticsearch-ci/docs

**Pull request merged**